File: lightroom-export-move.py
import os
from datetime import datetime

path = "C:\\Users\\noah\\AppData\\Local\\Temp\\Lightroom"
to_path = "C:\\Users\\noah\\Pictures\\Photography"

destination_directories = os.listdir(to_path)

for entry in os.scandir(path):
    if not entry.is_file() or not entry.name.endswith(".CR2"):
      continue

    date = datetime.fromtimestamp(entry.stat().st_mtime).date()

    # find destination directory that starts with the year
    destination_directory = next((x for x in destination_directories if x.startswith("{}-{}-{}".format(str(date.year), str(date.month), str(date.day)))), None)

    if destination_directory is None:
        print("Destination directory not found!")
        continue

    new_path = os.path.join(to_path, destination_directory, "Canon Rebel T7", "1. RAW")

    if not os.path.exists(new_path):
        print("Destination directory not found!")
        continue

    # move .xmp file
    xmp_file = entry.name.replace(".CR2", ".xmp")
    xmp_path = os.path.join(path, xmp_file)
    xmp_new_path = os.path.join(new_path, xmp_file)

    if not os.path.exists(xmp_path):
        print(".xmp not found!")
        continue

    print("Moving {} to {}".format(xmp_file, destination_directory))

    # delete existing .xmp file
    if os.path.exists(xmp_new_path):
        os.remove(xmp_new_path)

    os.rename(xmp_path, xmp_new_path)
    os.remove(entry.path)

File: lightroom-export-move.ts
import * as fs from "https://deno.land/std@0.167.0/fs/mod.ts";
import * as path from "https://deno.land/std@0.167.0/path/mod.ts";

const PATH = "C:\\Users\\noah\\AppData\\Local\\Temp\\Lightroom";
const TO_PATH = "C:\\Users\\noah\\Pictures\\Photography";

const start = Date.now();

const destination_directories: string[] = [];
for await (const entry of await Deno.readDir(TO_PATH)) {
  if (!entry.isDirectory) {
    continue;
  }

  destination_directories.push(entry.name);
}

let count = 0;
for await (const entry of await Deno.readDir(PATH)) {
  if (!entry.isFile || !entry.name.endsWith(".CR2")) {
    continue;
  }

  const cr2_path = path.join(PATH, entry.name);
  const date = (await Deno.stat(cr2_path)).mtime;

  if (!date) {
    console.log(`Could not find date for ${entry.name}`);
    continue;
  }

  const destination_directory = destination_directories.find((dir) =>
    dir.startsWith(`${date.getFullYear()}-${(date.getMonth() + 1).toString().padStart(2, "0")}-${date.getDate().toString().padStart(2, "0")}`)
  );

  if (!destination_directory) {
    console.log(`No destination directory found for ${entry.name}`);
    continue;
  }

  const destination_path = path.join(TO_PATH, destination_directory, "Canon Rebel T7", "1. RAW");
  if (!destination_path) {
    console.log(`No destination path found for ${entry.name}`);
    continue;
  }

  // move .xmp file
  const xmp_file = entry.name.replace(".CR2", ".xmp");
  const xmp_path = path.join(PATH, xmp_file);
  const xmp_destination_path = path.join(destination_path, xmp_file);

  if (!(await fs.exists(xmp_path))) {
    console.log(`No .xmp file found for ${entry.name}`);
    continue;
  }

  console.log(`Moving ${xmp_file} to ${xmp_destination_path}`);

  await fs.move(xmp_path, xmp_destination_path, {
    overwrite: true,
  });

  await Deno.remove(cr2_path);

  count++;
}

console.log(`Moved ${count} files in ${Date.now() - start}ms`);
